### DISCUSSION TITLE
Server rebinding to same address behaviour

### DISCUSSION THREAD

**Discussion by GrinningSoulGH** [hyperium/hyper#2636] [Q&A]
What I a trying to do is similar to this:
```
let server = Server::bind(&addr).serve(make_svc);
timeout(Duration::from_secs(1), server).await??;
Server::bind(&addr).serve(make_svc2).await?;
```
So, I want to rebind an address to a Server with a different request handler once an event occurs, a timeout in this example. I am aware of this approach: 
```
let server = Server::bind(&addr).serve(make_svc).with_graceful_shutdown(async { signal_receiver.await.ok(); });
sleep(Duration::from_secs(1)).await;
signal_sender.send(())?;
server.await?; // line in question
Server::bind(&addr).serve(make_svc2).await?;
```
How long can the commented line run for? Does it wait for all of request handler tasks to join? Can I let the hanging request handles run while migrating to a new Server? What I basically need is an ability to unbind the Server from the listening socket. What can I do with the current Hyper version that would be the best approximation of that?

**Comment by seanmonstar** [comment]
With graceful shutdown, once you trigger it, the listener will be stopped (so unbound), and it will be waiting for any spawned connections to finish. That's useful to know that the process can close cleanly without disrupting any connected clients.

If wanting to rebind that address, you can do so after triggering the graceful shutdown. You might not be able to do so immediately, as the kernel may reserve it for some time. You could also look into binding your own `TcpListener` and using REUSE_ADDR, so you can bind more quickly.

**Reply by GrinningSoulGH** [reply]
I tried that, and it fails on binding every time:
```
let (signal_sender, signal_receiver) = oneshot::channel();
let socket = TcpListener::bind(addr).await?;
let _server = Server::builder(AddrIncoming::from_listener(socket)?).serve(make_svc).with_graceful_shutdown(async { signal_receiver.await.ok(); });
signal_sender.send(())?;
sleep(Duration::from_secs_f32(10.0)).await;
let socket = TcpListener::bind(addr).await?;
let _server = Server::builder(AddrIncoming::from_listener(socket)?).serve(make_svc);
```
TcpListener is Tokio's, and it uses REUSE_ADDR. I need to await the server for it to work. Is there any way to make it possible to get the socket I passed into Server builder back during shutdown? That would solve the rebinding problem cleanly, in my opinion.

**Comment by rneswold** [comment]
Couldn't your `make_svc` function conditionally return a different service function based on whether the event occurred?
